### ISSUE TITLE
the problem with compiling tensorflow

### ISSUE THREAD

**Issue by mookor** [Nuitka/Nuitka#2862]
Hello!

I have a very simple one-line program
```
import tensorflow as tf
```
I am able to successfully assemble the executable file, but I get an error when starting it

```2024-05-15 20:57:03.973216: I external/local_tsl/tsl/cuda/cudart_stub.cc:32] Could not find cuda drivers on your machine, GPU will not be used.
2024-05-15 20:57:03.975392: I external/local_tsl/tsl/cuda/cudart_stub.cc:32] Could not find cuda drivers on your machine, GPU will not be used.
2024-05-15 20:57:04.002935: I tensorflow/core/platform/cpu_feature_guard.cc:210] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
[libprotobuf ERROR external/com_google_protobuf/src/google/protobuf/descriptor_database.cc:642] File already exists in database: tensorflow/core/protobuf/replay_log.proto
[libprotobuf FATAL external/com_google_protobuf/src/google/protobuf/descriptor.cc:1986] CHECK failed: GeneratedDatabase()->Add(encoded_file_descriptor, size): 
terminate called after throwing an instance of 'google::protobuf::FatalException'
  what():  CHECK failed: GeneratedDatabase()->Add(encoded_file_descriptor, size): 
Aborted (core dumped)
```

- Nuitka version, full Python version, flavor, OS, etc. as output by *this exact* command.

  > python -m nuitka --version
```
2.2.2
Commercial: None
Python: 3.10.14 (main, Apr  6 2024, 18:45:05) [GCC 9.4.0]
Flavor: Debian Python
Executable: /home/amazko/SNN2AltAI-Allocator/venv3.10/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu (based on Debian) 20.04.5
Version C compiler: /usr/bin/gcc (gcc 9).
```

- How did you install Nuitka and Python
 > pip install nuitka

installed in virtualenv

- The specific PyPI names and versions

  It should be taken from this output if there specific packages involved, e.g. `numpy`, you are
  expected to shorten this to the relevant ones.

  > python -m pip freeze
```
absl-py==2.1.0
astunparse==1.6.3
certifi==2024.2.2
charset-normalizer==3.3.2
flatbuffers==24.3.25
gast==0.5.4
google-pasta==0.2.0
grpcio==1.63.0
h5py==3.11.0
idna==3.7
keras==3.3.3
libclang==18.1.1
Markdown==3.6
markdown-it-py==3.0.0
MarkupSafe==2.1.5
mdurl==0.1.2
ml-dtypes==0.3.2
namex==0.0.8
Nuitka==2.2.2
numpy==1.26.4
opt-einsum==3.3.0
optree==0.11.0
ordered-set==4.1.0
packaging==24.0
protobuf==4.25.3
Pygments==2.18.0
requests==2.31.0
rich==13.7.1
six==1.16.0
tensorboard==2.16.2
tensorboard-data-server==0.7.2
tensorflow==2.16.1
tensorflow-io-gcs-filesystem==0.37.0
termcolor==2.4.0
typing_extensions==4.11.0
urllib3==2.2.1
Werkzeug==3.0.3
wrapt==1.16.0
zstandard==0.22.0
```

I tried to build the project like this:
```
python -m nuitka --standalone --include-package=tensorflow main.py
python -m nuitka --standalone  main.py
```

Nuitka log:
```
python -m nuitka --standalone main.py
Nuitka-Options: Used command line options: --standalone main.py
Nuitka: Starting Python compilation with Nuitka '2.2.2' on Python '3.10' commercial grade 'not installed'.
Nuitka-Plugins:anti-bloat: Not including '_json' automatically in order to avoid bloat, but this may cause: may slow down by using fallback implementation.
Nuitka: Completed Python level compilation and optimization.
Nuitka: Generating source code for C backend compiler.
Nuitka: Running data composer tool for optimal constant value handling.                                                                                
Nuitka: Running C compilation via Scons.
Nuitka-Scons: Backend C compiler: gcc (gcc 9).
Nuitka-Scons: Slow C compilation detected, used 360s so far, scalability problem.
Nuitka-Scons: Running gcc -o module.tensorflow.lite.python.schema_py_generated.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -fpartial-inlining -ftrack-macro-expansion=0 -Wno-deprecated-declarations
Nuitka-Scons: -fno-var-tracking -Wno-misleading-indentation -fcompare-debug-second -fno-lto -O2 -D__NUITKA_NO_ASSERT__ -D_NUITKA_STANDALONE -DPy_NO_ENABLE_SHARED -D_NUITKA_STATIC_LIBPYTHON
Nuitka-Scons: -D_NUITKA_USE_UNEXPOSED_API -D_NUITKA_CONSTANTS_FROM_INCBIN -D_NUITKA_FROZEN=151 -D_NUITKA_EXE -D_NUITKA_PLUGIN_MULTIPROCESSING_ENABLED=1
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/inline_copy/zlib -I/usr/include/python3.10 -I.
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/include -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/static_src
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/inline_copy/libbacktrace module.tensorflow.lite.python.schema_py_generated.c took 493.62 seconds
Nuitka-Scons: Slow C compilation detected, used 360s so far, scalability problem.
Nuitka-Scons: Slow C compilation detected, used 360s so far, scalability problem.
Nuitka-Scons: Running gcc -o module.tensorflow.python.ops.gen_math_ops.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -fpartial-inlining -ftrack-macro-expansion=0 -Wno-deprecated-declarations
Nuitka-Scons: -fno-var-tracking -Wno-misleading-indentation -fcompare-debug-second -fno-lto -O2 -D__NUITKA_NO_ASSERT__ -D_NUITKA_STANDALONE -DPy_NO_ENABLE_SHARED -D_NUITKA_STATIC_LIBPYTHON
Nuitka-Scons: -D_NUITKA_USE_UNEXPOSED_API -D_NUITKA_CONSTANTS_FROM_INCBIN -D_NUITKA_FROZEN=151 -D_NUITKA_EXE -D_NUITKA_PLUGIN_MULTIPROCESSING_ENABLED=1
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/inline_copy/zlib -I/usr/include/python3.10 -I.
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/include -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/static_src
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/inline_copy/libbacktrace module.tensorflow.python.ops.gen_math_ops.c took 372.64 seconds
Nuitka-Scons: Running gcc -o module.tensorflow.python.ops.gen_array_ops.o -c -std=c11 -fvisibility=hidden -fwrapv -pipe -fpartial-inlining -ftrack-macro-expansion=0 -Wno-deprecated-declarations
Nuitka-Scons: -fno-var-tracking -Wno-misleading-indentation -fcompare-debug-second -fno-lto -O2 -D__NUITKA_NO_ASSERT__ -D_NUITKA_STANDALONE -DPy_NO_ENABLE_SHARED -D_NUITKA_STATIC_LIBPYTHON
Nuitka-Scons: -D_NUITKA_USE_UNEXPOSED_API -D_NUITKA_CONSTANTS_FROM_INCBIN -D_NUITKA_FROZEN=151 -D_NUITKA_EXE -D_NUITKA_PLUGIN_MULTIPROCESSING_ENABLED=1
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/inline_copy/zlib -I/usr/include/python3.10 -I.
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/include -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/static_src
Nuitka-Scons: -I~/SNN2AltAI-Allocator/venv3.10/lib/python3.10/site-packages/nuitka/build/inline_copy/libbacktrace module.tensorflow.python.ops.gen_array_ops.c took 541.00 seconds
Nuitka-Scons: Backend linking program with 2720 files (no progress information available for this stage).
Nuitka-Scons:WARNING: You are not using ccache, re-compilation of identical code will be slower than necessary. Use your OS package manager to install it.
Nuitka-Plugins:dll-files: Found 1 file DLLs from tensorflow.lite.experimental.microfrontend.python.ops.audio_microfrontend_op installation.
Nuitka-Plugins:data-files: Included data file 'certifi/cacert.pem' due to package data for 'certifi'.                                             
Nuitka-Plugins:data-files: Included data file 'grpc/_cython/_credentials/roots.pem' due to package data for 'grpc'.
Nuitka: Keeping build directory 'main.build'.
Nuitka: Successfully created 'main.dist/main.bin'.
```

information about the system
```
lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.04.5 LTS
Release:        20.04
Codename:       focal
```

python version:
```
python --version
Python 3.10.14
```

gcc:
```
gcc --version
gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
```
P.S. I found a similar question here, but unfortunately I did not find an answer for myself there

**Comment by mookor** [comment]
I also tried to repeat this on another PC with a similar system + tried to use the dev version - it gives this error everywhere

**Comment by kayhayen** [comment]
I suspect this is tensorflow / protobuf in a newer version, I am trying to reproduce it now, but generally tensorflow was working previously, so this is a regression. Also I wonder what `tensorflow/core/protobuf/replay_log.proto` is, maybe it's a new data file, and that is causing this.

**Comment by mookor** [comment]
Good afternoon, thanks for the reply!

Could you tell me the versions of python/nuitka/tensorflow on which everything worked?

The strangest thing is that I cannot find this file on the PC, I tried searching through 
`sudo find / -name replay_log.proto`

**Comment by kayhayen** [comment]
I don't keep track of it like that. I saw the proto files in grep when using `--list-package-data=tensorflow`, I didn't check the precise file, but it seems there were really many of these.

**Comment by mookor** [comment]
also, i can compile without some dependencies 
python -m nuitka --standalone  --nofollow-import-to=tensorflow --nofollow-import-to=keras --nofollow-import-to=google main.py 

but if i run `./main.bin` 
it's show me error ModuleNotFoundError: No module named 'tensorflow'
How can I make the binary have a dependency on tensorflow from my environment?

**Comment by kayhayen** [comment]
You cannot add to standalone @mookor for that you would have to use accelerated mode, which has of course portability (to other systems) problems.

**Comment by Jzhenli** [comment]
@kayhayen I think nuitka may add a new compile option to skip compile 3rd None Pure-Python packages automatically. For the libs such as tensorflow, the lib owner have compiled them to pyd/so files. Nuitka should focus on the Pure-Python packages.

**Comment by kayhayen** [comment]
In the plugins, you can decide the compilation mode and, if at all, on a per-module basis; however, if you put "bytecode" there, it will not necessarily work at all. One typical upcoming example will be that the source code will be needed for JITs or whatever, and I don't really want to find out myself why then things don't work and support that. By the time the bytecode is created in a good way and is re-used cached, then it can also just generate the C code and be good too, since that is not really the problem usually.

The real solution is improved scalability of Nuitka, which will be my focus in the summer. With a 10x improvement in Python compile time, and maybe caching of its result one day, we will be good. Also with 3.13, we might try and compile with the no-GIL variant eventually in parallel or we use multiprocessing, but that will require caching to work, because it's going to be how the result is communicated back.

Anyway, it's a nice idea, and if you can write a plugin that makes it work for you, and you feel that is time spent well, go on. For me, it's a diversion from my main goals.

**Comment by kayhayen** [comment]
So, I reproduced it and found that the string in question lives in `MiniTensorflowOpt.dist/tensorflow/python/_pywrap_tensorflow_internal.so`. Of course, the question now is why, if at all, it is loaded twice, this may just misdiagnosis of that code. My googling found only cases of unsupported platforms.

**Comment by kayhayen** [comment]
So, this is probably the issue:

```
grep: MiniTensorflowOpt.dist/tensorflow/python/_pywrap_tensorflow_internal.so: binary file matches
grep: MiniTensorflowOpt.dist/_pywrap_tensorflow_internal.so: binary file matches
```

This is because of those:

```
<included_extension name="_pywrap_tensorflow_internal.so" dest_path="tensorflow/python/_pywrap_tensorflow_internal.so" source_path="~/.local/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so" package="tensorflow.python" ignored="no" reason="used extension module" />
<included_dll name="_pywrap_tensorflow_internal.so" dest_path="_pywrap_tensorflow_internal.so" source_path="~/.local/lib/python3.9/site-packages/tensorflow/python/_pywrap_tensorflow_internal.so" package="tensorflow.compiler.mlir.quantization.tensorflow.calibrator" ignored="no" reason="Used by 'tensorflow/compiler/mlir/quantization/tensorflow/calibrator/pywrap_calibration.so'" />
```

So, it's found as both an extension module and a DLL, which is very strange.

**Comment by kayhayen** [comment]
One thing I noticed and dislike, is this:

```
        libtensorflow_framework.so.2 => /home/hayen/repos/Py2C/MiniTensorflowOpt.dist/tensorflow/compiler/mlir/quantization/tensorflow/calibrator/../../../../../../libtensorflow_framework.so.2 (0x00007f0d950df000)
```

It seems, that it's carried over from the original PyPI package

```
        libtensorflow_framework.so.2 => /home/hayen/.local/lib/python3.9/site-packages/tensorflow/compiler/mlir/quantization/tensorflow/calibrator/../../../../../libtensorflow_framework.so.2 
```

Of course, that is very ugly and may, in fact, cause issues if those directories do not all exist, although I think that at least Linux is forgiving there, so that's not our issue.

What happens, though, it seems, is that there is a dependency as a DLL on an extension module.

```
        _pywrap_tensorflow_internal.so => /home/hayen/.local/lib/python3.9/site-packages/tensorflow/compiler/mlir/quantization/tensorflow/calibrator/../../../../../python/_pywrap_tensorflow_internal.so 
```

Nuitka places DLLs near the main binary by default to make it most likely to be found by everything.

**Comment by kayhayen** [comment]
So, I guess the solution is to remove DLLs from detection, which are just extension modules and hope for the best, namely that they are found.

**Comment by kayhayen** [comment]
I have done this, but I am actually uncertain about the current state of this issue now.

**Comment by kayhayen** [comment]
This is resolved, at least with 2.5 probably long before already, but I didn't trace it properly.
